File: 04 - Binary Tree - Divide Conquer & Traverse/E453.FlattenBinaryTreetoLinkedList.py
"""
Definition of TreeNode:
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left, self.right = None, None
"""

class Solution:
    last_node = None
    
    """
    @param root: a TreeNode, the root of the binary tree
    @return: nothing
    """
    def flatten(self, root):
        # write your code here

        # method 2:
        if root is None:
            return None

        if self.last_node is not None:
            self.last_node.left = None
            self.last_node.right = root

        self.last_node, right_node = root, root.right

        self.flatten(root.left)
        self.flatten(right_node)

    #     # method 1:
    #     self.helper(root)
    #
    # def helper(self, root):
    #     if root is None:
    #         return None
    #
    #     print(root.val)
    #     left_last, right_last = self.helper(root.left), self.helper(root.right)
    #
    #     if left_last is not None:
    #         left_last.right, root.right, root.left = root.right, root.left, None
    #
    #     if right_last is not None:
    #         return right_last
    #
    #     if left_last is not None:
    #         return left_last
    #
    #     return root

File: 04 - Binary Tree - Divide Conquer & Traverse/E453.FlattenBinaryTreetoLinkedList.java
/**
 * Definition of TreeNode:
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left, right;
 *     public TreeNode(int val) {
 *         this.val = val;
 *         this.left = this.right = null;
 *     }
 * }
 */

public class Solution {
    // private TreeNode lastNode = null;
    /**
     * @param root: a TreeNode, the root of the binary tree
     * @return: nothing
     */
    public void flatten(TreeNode root) {
        // write your code here
        //if (root == null) return;

        // method 2:
        // flatten(root.right);
        // flatten(root.left);

        // root.right = lastNode;
        // root.left = null;
        // lastNode = root;

        // method 1:
        // if (lastNode != null) {
        //     lastNode.left = null;
        //     lastNode.right = root;
        // }

        // lastNode = root;
        // TreeNode rightNode = root.right;

        // flatten(root.left);
        // flatten(rightNode);

        // method 3:
        helper(root);
    }

    // flattern root and all its children nodes into a list
    // return the list's head and tail.
    private TreeNode[] helper(TreeNode root) {
        if (root == null) return null;

        TreeNode[] leftNode = helper(root.left);
        TreeNode[] rightNode = helper(root.right);

        root.left = null;
        TreeNode tail = root;

        if (leftNode != null) {
            tail.right = leftNode[0];
            tail = leftNode[1];
        }
        if (rightNode != null) {
            tail.right = rightNode[0];
            tail = rightNode[1];
        }

        return new TreeNode[]{root, tail};
    }
}
